src/bucklescript/output_bucklescript_types.re

```diff
@@ -0,0 +1,419 @@
+  config.records || force_record
+    ? generate_record_type(config, fields, obj_path)
+    : generate_object_type(config, fields, obj_path);
+};
+
+// generate all the types necessary types that we later refer to by name.
+let generate_types = (config: Generator_utils.output_config, res) => {
+  let types =
+    extract([], res)
+    |> List.map(
+         fun
+         | Object({fields, path: obj_path, force_record}) =>
+           generate_graphql_object(config, fields, obj_path, force_record),
+       );
+
+  Ast_helper.Str.type_(Recursive, types);
+};
+
+let rec generate_arg_type = loc =>
+  fun
+  | Type(Scalar({sm_name: "ID"}))
+  | Type(Scalar({sm_name: "String"})) => base_type("string")
+  | Type(Scalar({sm_name: "Int"})) => base_type("int")
+  | Type(Scalar({sm_name: "Float"})) => base_type("float")
+  | Type(Scalar({sm_name: "Boolean"})) => base_type("bool")
+  | Type(Scalar({sm_name: _})) => base_type("Js.Json.t")
+  | Type(Enum(enum_meta)) =>
+    Graphql_ppx_base__.Schema.(
+      Ast_helper.(
+        Typ.variant(
+          enum_meta.em_values
+          |> List.map(({evm_name, _}) =>
+               Rtag({txt: evm_name, loc: Location.none}, [], true, [])
+             ),
+          Closed,
+          None,
+        )
+      )
+    )
+  | Type(InputObject({iom_name})) =>
+    base_type(generate_type_name(~prefix="t_variables", [iom_name]))
+  | Type(Object(_)) =>
+    raise(
+      Location.Error(
+        Location.error(~loc=loc |> conv_loc, "Object not allowed in args"),
+      ),
+    )
+  | Type(Union(_)) =>
+    raise(
+      Location.Error(
+        Location.error(~loc=loc |> conv_loc, "Union not allowed in args"),
+      ),
+    )
+  | Type(Interface(_)) =>
+    raise(
+      Location.Error(
+        Location.error(~loc=loc |> conv_loc, "Interface not allowed in args"),
+      ),
+    )
+  | Nullable(inner) =>
+    base_type(~inner=[generate_arg_type(loc, inner)], "option")
+  | List(inner) =>
+    base_type(~inner=[generate_arg_type(loc, inner)], "array")
+  | TypeNotFound(name) =>
+    raise(
+      Location.Error(
+        Location.error(
+          ~loc=loc |> conv_loc,
+          "Type " ++ name ++ " not found!",
+        ),
+      ),
+    );
+
+let generate_record_input_object = (input_obj_name, fields) => {
+  Ast_helper.Type.mk(
+    ~kind=
+      Ptype_record(
+        fields
+        |> List.map(
+             fun
+             | InputField({name, type_, loc}) => {
+                 Ast_helper.Type.field(
+                   {Location.txt: name, loc: Location.none},
+                   generate_arg_type(loc, type_),
+                 );
+               },
+           ),
+      ),
+    {
+      loc: Location.none,
+      txt:
+        generate_type_name(
+          ~prefix="t_variables",
+          switch (input_obj_name) {
+          | None => []
+          | Some(name) => [name]
+          },
+        ),
+    },
+  );
+};
+
+let generate_object_input_object = (input_obj_name, fields) => {
```

Comment:
Would it be possible to deduplicate logic inside `generate_object_input_object` and `generate_record_input_object` ?

Reply:
The code is quite different, not really duplicated logic, or do I miss something?